design/ratelimit-design-new.md

```diff
@@ -0,0 +1,180 @@
+If no global RLS is defined, no global rate limiting will occur.
+
+## Detailed Design
+
+### RateLimitPolicy type
+
+A sample `RateLimitPolicy` looks like:
+
+```yaml
+rateLimitPolicy:
+  # local defines local rate limiting properties for the virtual host or route.
+  local:
+    # tokenBucket defines the properties of the local rate limiting token bucket.
+    # ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket
+    tokenBucket:
+      maxTokens: 100
+      tokensPerFill: 10
+      fillInterval: 1s
+  # global defines global rate limiting properties for the virtual host or route.
+  global:
+    # rateLimitDescriptors defines the descriptors to be generated and sent to the
+    # external Rate Limiting Service (RLS) for a rate limit decision.
+    rateLimitDescriptors:
+    # This descriptor is generated only if the x-steve-ratelimit header
+    # is present on the request.
+    - items:
+        # adds a descriptor entry with a key of "generic_key" and a value
+        # of "s1".
+        - genericKey:
+            value: s1
+
+        # adds a descriptor entry with a key of "remote_address" and a value
+        # equal to the trusted address from x-forwarded-for.
+        - remoteAddress: {}
+
+        # adds a descriptor entry with a key of "steve-ratelimit"
+        # and a value equal to the value of the "x-steve-ratelimit" header.
+        - requestHeader:
+            descriptorKey: steve-ratelimit
+            headerName: x-steve-ratelimit
+        
+    # This descriptor is always generated since it's just a static value.
+    - items:
+        - genericKey:
+            value: s1
+```
+
+This API maps closely to Envoy's rate limiting API.
+
+For local rate limiting, the rate limit itself is defined inline in the `RateLimitPolicy`, as token bucket properties.
+Since local rate limits apply *per Envoy*, each pod in the Envoy daemon set will have a token bucket with these properties.
+
+For global rate limiting, per Envoy's API, the `RateLimitPolicy` only defines a list of descriptors to send to an external RLS for a given request.
+All [descriptor entries defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action) will be supported except `metadata` and `dynamic_metadata`.
+
+### Contour Configuration
+If using global rate limiting, the details of the external RLS must be configured with Contour.
+First, an `ExtensionService` must be defined for the RLS with the cluster-level details for the RLS itself.
+For example:
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: ratelimit
+spec:
+  protocol: h2
+  services:
+    - name: ratelimit
+      port: 8081
+  timeoutPolicy:
+    response: 50ms
```

Comment:
There's also a `timeout` property of the rate_limit service (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ratelimit/v3/rate_limit.proto) , should we expose that as well?

Reply:
Just reading the comments on the struct it wasn't clear to me.